lib/auth/auth_with_roles.go

```diff
@@ -218,13 +218,100 @@ func (a *ServerWithRoles) actionWithExtendedContext(ctx context.Context, kind, v
 func (a *ServerWithRoles) actionForKindSession(ctx context.Context, sid session.ID) error {
 	extendContext := func(servicesCtx *services.Context) error {
 		sessionEnd, err := a.findSessionEndEvent(ctx, sid)
+		if err != nil {
+			return trace.Wrap(err)
+		}
 		servicesCtx.Session = sessionEnd
+		servicesCtx.Resource = rebuildResourceFromSessionEndEvent(sessionEnd)
+		// Also add the roles of the user to the context so that they can be used
+		// in the `where` by `can_view` function.
+		servicesCtx.Roles = a.context.Checker.Roles()
+		return nil
+	}
+
+	// First try a simple check without the extended context.
+	if err := a.actionWithContext(&services.Context{User: a.context.User}, types.KindSession, types.VerbRead); err == nil {
+		return nil
+	} else if !trace.IsAccessDenied(err) {
+		// If the error is something other than AccessDenied, return it.
 		return trace.Wrap(err)
 	}
 
 	return trace.Wrap(a.actionWithExtendedContext(ctx, types.KindSession, types.VerbRead, extendContext))
 }
 
+// rebuildResourceFromSessionEndEvent rebuilds a resource from a session end event.
+// This is used to reconstruct the resource that was active at the time of the session end event
+// for audit log RBAC purposes.
+func rebuildResourceFromSessionEndEvent(event apievents.AuditEvent) types.Resource {
```

Comment:
This seems like an important thing to test, but I don't see any test coverage in lib/auth for it.